Add metadata context support to eliminate redundant metadata storage

xiaguan opened the pull request:
## Summary

This PR introduces metadata context support to connectors, eliminating the need to store redundant metadata with each key in remote storage and simplifying the data flow.

## Changes

- **Add ConnectorMetadataContext class**: Provides shape, dtype, and format information for KV cache tensors
- **Update MooncakeStore connector**: Remove metadata serialization/deserialization, use context-provided metadata instead
- **Add metadata context delegation**: Ensure wrapper connectors (audit, instrumented) properly delegate metadata context to underlying connectors
- **Set up metadata context in RemoteBackend**: Automatically configure metadata context based on engine metadata

## Benefits

- **Reduced storage overhead**: No longer store identical metadata with each key
- **Simplified data flow**: Remove metadata serialization/deserialization steps
- **Better performance**: Eliminate redundant metadata operations

xiaguan commented:
Hey @YaoJiayi , mind taking a look? This change would let us pre-allocate the buffer, which means we could optimize for zero-copy in the future.

xiaguan commented on lmcache/v1/storage_backend/connector/mooncakestore_connector.py:
```diff
@@ -103,16 +101,14 @@ def __init__(
             logger.info("Mooncake Configuration loaded. config: %s", self.config)
 
-            self.store.setup_lmcache(
+            self.store.setup(
```
I should take forward compatibility into account. I'll make some updates here.

cicirori commented:
Thanks for your work! I noticed in local testing that this PR changes some of the existing behavior.

**Before:**

```
Total tokens 10011, LMCache hit tokens: 10010, need to load: 10010 (vllm_v1_adapter.py:745:lmcache.integration.vllm.vllm_v1_adapter)
```

**After applying the PR:**

```
Total tokens 10011, LMCache hit tokens: 9984, need to load: 9984 (vllm_v1_adapter.py:745:lmcache.integration.vllm.vllm_v1_adapter)
Storing KV cache for 27 out of 10011 tokens (skip_leading_tokens=9984) for request 58
(vllm_v1_adapter.py:688:lmcache.integration.vllm.vllm_v1_adapter)
```

I wanna know if this change in token accounting is expected.

xiaguan commented:
expected.since we use static metadata with full chunks (e.g., the default is 256), so we do not store incomplete chunks.

Perhaps we should add a check to ensure that Mooncake is used with the `save_unfull_chunk` flag enabled—this could lead to more consistent behavior.

xiaguan commented:
Closing this PR for now — I’ll open a new one that adds support for unfull chunks.

xiaguan committed:
Fix formatting: remove trailing whitespace

Signed-off-by: susun <751080330@qq.com>

xiaguan committed:
Add metadata context support to connectors

- Add ConnectorMetadataContext class to base_connector.py
- Update mooncakestore_connector to use metadata context instead of storing metadata with each key
- Add metadata context delegation to audit_connector and instrumented_connector
- Set up metadata context in remote_backend based on engine metadata
- Remove metadata serialization/deserialization from mooncake operations

Signed-off-by: Jinyang Su <751080330@qq.com>
Signed-off-by: susun <751080330@qq.com>

xiaguan committed:
feat: improve MooncakeStore connector with better unfull chunk handling

- Remove initialization check that required save_unfull_chunk to be disabled
- Add proper memory allocation before data retrieval
- Implement robust unfull chunk handling with proper reshaping
- Add comprehensive error checking for buffer size alignment
- Support KV_2LTD format with dynamic sequence length adjustment
- Add assertions to validate memory object metadata matches context
- Improve error messages and logging for better debugging

Signed-off-by: susun <751080330@qq.com>

xiaguan committed:
Update mooncakestore_connector.py - fix line length formatting

Signed-off-by: Jinyang Su <751080330@qq.com>

xiaguan commented:
@maobaolong PTAL，it works fine in my env

xiaguan committed:
feat(connector): enable zero-copy operations with mooncake store

Signed-off-by: Jinyang Su <751080330@qq.com>

maobaolong commented:
@chunxiaozheng Would you like to take a look at this PR?

LuyuZhang00 commented:
If this PR will be merged later?

xiaguan commented:
duplicate with https://github.com/LMCache/LMCache/pull/1269 close
